### PR TITLE
Extend decimal point and Misc.

### PR DESCRIPTION
**Pull request by tirix**
In this PR, I'm pulling the chapter "Decimal Point" from David's Mathbox into mine, and extend it with a few theorems.

Tough, I'd like to propose a slightly different construct for decimal point:
My proposition is to add a marker _after_ the decimal expansion, like in ` ( 3 . 1 4 1 5 9 : : : : ) ` . 
I believe this is a bit more readable than the original proposal which would result in interleaving underscores in the decimal expansion, as in ` ( 3 . _ 1 _ 4 _ 1 _ 5 9 ) ` .

I've kept both constructions (previous one with underscore as `df-dp2` , and new one as `dp-dp3`), and I've also kept the attributions to David.

Apart from a few generic theorems moved from other people's mathboxes to the main part, the other changes are within my mathbox.

### CODE REVIEW

**Comment by tirix** [comment]
It looks like the CI does not fail because of any issue with discouraged theorems, but rather because the `discouraged` file could not be built:
```
Run echo 'Checking discouraged file:' && diff -U 0 discouraged discouraged.new
  echo 'Checking discouraged file:' && diff -U 0 discouraged discouraged.new
  shell: /usr/bin/bash -e {0}
  env:
    JAVA_HOME_10.0.2_x64: /opt/hostedtoolcache/jdk/10.0.2/x64
    JAVA_HOME: /opt/hostedtoolcache/jdk/10.0.2/x64
    JAVA_HOME_10_0_2_X64: /opt/hostedtoolcache/jdk/10.0.2/x64
Checking discouraged file:
diff: discouraged.new: No such file or directory
Error: Process completed with exit code 2.
```
Actually, MMJ2 crashes when generating the file:
```
java.lang.NullPointerException
	at mmj.verify.LRParser.initialize(LRParser.java:175)
	at mmj.verify.LRParser.load(LRParser.java:320)
	at mmj.verify.LRParser.load(LRParser.java:317)
	at mmj.verify.LRParser.parseExpr(LRParser.java:382)
	at mmj.verify.Grammar.grammaticalParseSyntaxExpr(Grammar.java:732)
	at mmj.verify.GrammarAmbiguity.basicAmbiguityEdits(GrammarAmbiguity.java:208)
	at mmj.verify.Grammar.initializeGrammarTables(Grammar.java:1008)
	at mmj.verify.Grammar.initializeGrammar(Grammar.java:704)
	at mmj.util.GrammarBoss.initializeGrammar(GrammarBoss.java:288)
	at mmj.util.GrammarBoss.doParse(GrammarBoss.java:242)
	at mmj.util.Boss.lambda$0(Boss.java:107)
	at mmj.util.Boss.doRunParmCommand(Boss.java:121)
	at mmj.util.BatchFramework.executeRunParmCommand(BatchFramework.java:281)
	at mmj.util.BatchFramework.runIt(BatchFramework.java:223)
	at mmj.util.BatchMMJ2.main(BatchMMJ2.java:53)
```
I regenerated it locally, but with the Metamath program.

**Comment by benjub** [comment]
Wouldn't this make the grammar more complicated and the file harder to parse ?  Are there other cases in set.mm of n-ary postfix notation with n>1 ?  It seems that it would introduce the need of unbounded look-ahead.

**Comment by tirix** [comment]
That might be the reason, but actually all (other) verifiers are fine with it, and I've used "E-Metamath" to develop it, which is based on MMJ2, so I'm puzzled why this would happen for generating "discouraged" and not for the rest. 
I'd like to take time to review and step-through MMJ2 to fully understand the issue, but I did not have the opportunity yet.

**Comment by digama0** [comment]
I'm glad the LRParser caught this. This is indeed not an allowed grammar construction, because it violates condition (ii) of theorem 4 in http://us.metamath.org/downloads/grammar-ambiguity.txt : no proper prefix of a class expression is a class expression. This can lead to ambiguous parses.

The reason it was not caught by other verifiers is that grammar ambiguity is difficult to check and not required by the metamath spec, but it can lead to misleading theorems. Most metamath verifiers / proof assistants either don't do math parsing at all or parse dynamically, which means that unambiguity is only checked if it actually occurs in the statements that appear in the database (and sometimes not even then; many parsers just take some parse at random). LRParser is the only existing metamath parser that can check unambiguity up-front, which is part of why it is in CI.

For a concrete example of an ambiguous expression, consider `; A B C :`, which can be either `cdc cA (cdp3 cB cC)` or `cdp3 (cdc cA cB) cC`. This one is "benign" since the two expressions are provably equal, but `-u A B :` has the same issue and is not benign.

**Comment by david-a-wheeler** [comment]
Yes, I agree with @digama0 - we prevent ambiguous parses, by ensuring that constructs can be unambiguously resolved after a certain number of tokens (I believe the number is 5 in parse system). Which is why I defined `_` in this way. Alternatives are worth considering, but I don't think creating an ambiguous parse is worth it.

**Comment by david-a-wheeler** [comment]
BTW, just to make it clear, you are *absolutely* welcome to move decimal point from my mathbox to yours & extend it. I think it'd be good to have *some* support for the decimal point in the main body; I hear they are used sometimes :-).

**Comment by tirix** [comment]
Thank you Mario for the explanation! 

I would not say that the LRParser caught it, or at least it did not catch it gracefully, since it crashed without clear message: maybe we could improve the error handling / error message? (interestingly, metamath-knife also panicked without any explanation, so there is also room for improvement there!)

I'll see what I can do!

**Comment by tirix** [comment]
Thanks Mario for the additional explanation, especially the example of ambiguous expression: this proves that this construct is ambiguous.

And thanks David for allowing me to experiment with the section you have initiated!

**TL; DR: The last commit reverts to David's original construct and passes all checks. I propose to first move on with it.**

I think there is some value in keeping the decimal expansion uninterrupted, in order to stay near to the usual decimal representation, which is something Norm valued.

I'm taking the opportunity to list my thoughts about different alternatives:
Reminder: With the current solution, two examples of expressions representing decimal numbers would be
` ( 3 . _ 1 _ 4 _ 1 _ 5 9 ) ` and ` ( ; ; 1 0 0 . _ 0 _ 0 1 ) ` .

### Solution 1: A new type code
I believe I have seen somewhere Norm or Mario tried this without success but can't find where anymore. A new type code would allow to define completely new parsing constraint, including a suffix notation. However, it would break the beautifully minimal wff/set/class triplet, and is way too heavy-handed for this simple target.

### Solution 2: Using 3 new constructs
Since we can only group uninterrupted list of tokens left-to right, for decimals, we have to know what place is the next one. One could use pairs to encode that, with the first element of the pair being the current value, and the second being the current place. This would require 3 constructs:

* an initial one to construct the first pair: e.g. ` A . B `
* an iteration step which adds a number to the decimal expansion: ` : A B `
* a final step which unpacks the pair and drops the second element: ` ." A ". `

A resulting decimal expression would be something like ` ." : : : : 3 . 1 4 1 5 9 ". ` or ` ." : : ; ; 1 0 0 . 0 0 1 ". `
Here is what it might look like:
```
$c ." ". . $.
$a class A . B $.
$a class : A B $.
$a class  ." A ". $.
$a |- A . B = <. ( A + ( B / ; 1 0 ) ) , 2 .> $.
$a |- : A B = <. ( ( 1st ` A ) + ( B x. ( ; 1 0 ^ -u ( 2nd ` A ) ) ) ) , ( ( 2nd ` A ) + 1 ) .> $.
$a |- ." A ". = ( 1st ` A ) $.
```
I believe this is too convoluted, and hard to understand.

### Solution 3: Using words
Word constructs have the nice property that the string of tokens is uninterrupted. With this definition, and no new construct:
```
$a |- . = ( a e. NN0 , b e. Word NN0 |-> ( a + sum_ i e. dom b ( ( b ` i ) x. ( ; 1 0 ^ -u ( i + 1 ) ) ) ) $.
```
We could write decimal numbers as ` ( 3 . <" 1 4 1 5 9 "> ) ` or ` ( ; ; 1 0 0 . <" 0 0 1 "> ) `
Words are currently limited to 8 places, but this would cover most usages. The advantage here is that we don't need to define any new syntax construct. Calculus might however be unnecessarily complicated.

### Solution 4: Word-like constructs
One issue with the previous syntax is that the fist ` <" ` token comes in the middle of the decimal expression. One could define a series of constructs of different lengths, just like for words, for decimal expansions of different lengths.
```
$c . ." ". $.
$a class ." A . B ". $.
$a class ." A . B C ". $.
$a class ." A . B C D ". $.
etc.
```
I think the issue here is that when moving to doing calculus, these may then require an onslaught of theorems in order to cover all cases, or conversion theorems to move from one form to another one. But the advantage here is that the final syntax would be very readable: ` ." 3 . 1 4 1 5 9 ". ` which is my graal here, or ` ." ; ; 1 0 0 . 0 0 1 ". `. 

One may then wish to extend this to also get rid of the initial ` ; ` notation, allowing for example to write ` ." 1 0 0 . 0 0 1 ". `. Unfortunately the number of cases then explodes quadratically if we are to allow any number of places before and after the dot.

In this case it is interesting to use this kind of definition:
```
$a |- ." A . B C D ". = ( A + ( ." B . C D ". / ; 1 0 ) ) $. 
```
rather than something like:
```
$a |- ." A . B C D ". = ( ." A . B C ". + ( D / ; ; ; 1 0 0 0 ) ) $. 
```
Because then, all definitions have the same form ` ( A + ( B / ; 1 0 ) ) ` , which can become a common essential hypothesis and lead to common generic theorems for all syntax forms. This sounds like an interesting idea and I might experiment with it in the future, but for the moment I'm rather interested in the application than the actual decimal expansions, and I'd like to move on with the current definition.

**Comment by digama0** [comment]
Personally, I have found decimal numbers to be of very limited use, and I don't think that any of the proposed alternatives are better than David's solution, but I would suggest rather than any of them to simply use rational expressions like `( 2 5 3 ; ; / 3 5 7 ; ; )`, which are much simpler to work with in all respects, and can be used for much more accurate approximations than decimals admit. (More precisely, you generally need numbers with half as many digits to represent a random real to a given precision, compared to decimals; multiplications on such numbers are therefore 4x smaller and faster.)

**Comment by tirix** [comment]
Yes. Here, I specifically wanted to stick as much as possible to the way the theorems were expressed in the original article, and they were using decimal expansions.

You can find these theorems further down in this PR's change set, some of them still commented out.

**Comment by david-a-wheeler** [comment]
@tirix - merge conflicts!

**Comment by david-a-wheeler** [comment]
Oh, and "fration" should be "fraction". That's my fault, but since you're there you can fix it :-).

**Comment by tirix** [comment]
Merge conflict is resolved and typo is fixed.
I think discussions are closed so this would be ready to be merged!

File: set.mm [hunk 1]
```diff
@@ -489156,6 +489830,362 @@ of the indicator function (Contributed by Thierry Arnoux,
...
+      ad2antlr r19.21bi simpr2 simpr3 tgoldbachgtda hgt749d r19.29vva ) AHJZUAJ
+      ZKUCKLMUDUHUIUINNNNUEOUFPZHUGUJZWFUBJZKUCKLUKLUKNNUEOUFPZHUGUJZMMMMUKQQUK
+      ULNNNNNNNUEOCQUMOROIMLUNOIJZUOWJRUPZOCUQOKWMUOWGWNOCUQOKQUMOROURQUSROROCU
+      TWMROROVAKROVFUFPZVBZMDVCVDCVEVGKOVHKVIPUBUASTVJOZWQAWJWQVKZVLZWGWQVKZVLZ
+      WPVLZIBHWJWGCDEACDVKWRWTWPFVMALMVNQUDVNUMOCUFPWRWTWPGVMWRTSWJVOAWTWPWJSTV
+      PVQWTTSWGVOWSWPWGSTVPVSXBWHHUGXAWIWLWOVRVTXBWKHUGXAWIWLWOWAVTXAWIWLWOWBWC
+      AIBUBUAHCDEFGWDWE @.
+  @}
+$)
+
```
Comment (avekens) [inline set.mm:490188]: That would be really great if you can prove this! It corresponds to ~ax-tgoldbachgt in my mathbox, which I provided as axiom because I coud not prove it. As soon as you have finished your work, we should talk about merging our results, and maybe move them to the main body of set.mm. Of course we have to harmonize/discuss the definition of odd numbers (my class ` Odd ` and your class ` O = { z e. ZZ | -. 2 || z } `, see discussion in #1682).

**Review by avekens — APPROVED** [review]

File: set.mm [hunk 2]
```diff
@@ -465654,6 +465751,432 @@ its graph has a given second element (that is, function value).
...
+    $( Equality theorem for the decimal expansion constructor.  (Contributed by
+       David A. Wheeler, 15-May-2015.) $)
+    dp2eq12i $p |- _ A C = _ B D $=
+      ( cdp2 dp2eq1i dp2eq2i eqtri ) ACGBCGBDGABCEHCDBFIJ $.
+  $}
+
+  ${
+    dp20u.1 $e |- A e. NN0 $.
+    $( Add a zero in the tenths (lower) place.  (Contributed by Thierry Arnoux,
+       16-Dec-2021.) $)
```
Comment (benjub) [inline set.mm:465822]: tenths ?
"Adding a zero as the least significant digit after a decimal separator does not change the value of a number." ?

File: set.mm [hunk 3]
```diff
@@ -465654,6 +465751,432 @@ its graph has a given second element (that is, function value).
...
+    $( Add a zero in the unit places.  (Contributed by Thierry Arnoux,
+       16-Dec-2021.) $)
+    dp20h $p |- _ 0 A = ( A / ; 1 0 ) $=
+      ( cc0 cdp2 c1 cdc cdiv co caddc df-dp2 crp wcel cc ax-mp 10nn0 nn0cni 0re
+      rpcn 10pos gtneii divcli addid2i eqtri ) CADCAECFZGHZIHUECAJUEAUDAKLAMLBA
+      RNUDOPCUDQSTUAUBUC $.
+  $}
+
+  $( Define the closure for the decimal fraction constructor if both values are
+     reals.  (Contributed by David A. Wheeler, 15-May-2015.) $)
```
Comment (benjub) [inline set.mm:465840]: Define the closure for --> Closure of/for
?

File: set.mm [hunk 4]
```diff
@@ -465654,6 +465751,432 @@ its graph has a given second element (that is, function value).
...
+       ` -u ( ; 3 2 . _ 7 _ 1 8 ) = -u ( ; ; ; ; 3 2 7 1 8 / ; ; ; 1 0 0 0 ) `
+       Unary minus, if applied, should normally be applied in front of the
+       parentheses.
+
+       Metamath intentionally does not have a built-in construct for numbers,
+       so it can show that numbers are something you can build based on set
+       theory.  However, that means that metamath has no built-in way to handle
+       decimal numbers as traditionally written, e.g., "2.54", and its parsing
+       system intentionally does not include the complexities necessary to
+       define such a parsing system.  Here we create a system for modeling
```
Comment (benjub) [inline set.mm:465959]: "parsing system" twice ?

File: set.mm [hunk 5]
```diff
@@ -465654,6 +465751,432 @@ its graph has a given second element (that is, function value).
...
+
+  Define the decimal point operator and the decimal fraction constructor.
+  This can model traditional decimal point notation, and serve
+  as a convenient way to write some fractional numbers.
+  See ~ df-dp and ~ df-dp2 for more information; ~~? dfpval
+  provides a more convenient way to obtain a value.
+  This is intentionally similar to ~ df-dec .
+
+  TODO: Fix non-existent label dfpval.
+
```
Comment (benjub) [inline set.mm:465937]: Is there a dfpval to be added, or is it simply dpval ? (in which case, you may remove the TODO and the "~~?" ).
You may also rewrap this to width 79, and have a double blank line before that subsection.

**Review by benjub — APPROVED** [review]
I made some comments that you may follow (or not) in a later PR.

File: set.mm (hunk 2 above)
Comment (tirix) [reply set.mm:465822]: Yes, the tenths place! 
See for example [https://en.wikipedia.org/wiki/Numerical_digit#Computation_of_place_values](the Wikipedia for the place value). In french we might say "les unités, les dixièmes, les centièmes"

File: set.mm (hunk 3 above)
Comment (tirix) [reply set.mm:465840]: Agreed! This will be fixed in my next commit.

File: set.mm (hunk 4 above)
Comment (tirix) [reply set.mm:465959]: This is originally for David, but I agree with your comment, and it will be changed it in my next commit.

File: set.mm (hunk 5 above)
Comment (tirix) [reply set.mm:465937]: This is also from @david-a-wheeler . In the line above, there is a reference to a ~ dpfrac theorem which had not been proven yet.  In my next commit, I'm removing this TODO and I refer to ~ dfpval2 and ~ dpfrac1.

**Comment by david-a-wheeler** [comment]
@benjub - feel free to fix it up as you think best. I'm just glad someone else also thinks it has some utility!
